formats/json-tests/commonTest/src/kotlinx/serialization/json/JsonElementDecodingTest.kt

```diff
@@ -51,4 +53,58 @@ class JsonElementDecodingTest : JsonTestBase() {
         json = json.replace("%", "0")
         Json.parseToJsonElement(json)
     }
+
+    private open class NullableSerializer<T : Any>(private val serializer: KSerializer<T>) : KSerializer<T?> {
+        final override val descriptor: SerialDescriptor = serializer.descriptor.nullable
+
+        final override fun serialize(encoder: Encoder, value: T?) {
+            if (value == null) {
+                encoder.encodeNull()
+            } else {
+                serializer.serialize(encoder, value)
+            }
+        }
+
+        final override fun deserialize(decoder: Decoder): T = serializer.deserialize(decoder)
```

Comment:
Incorrect, must call `decodeNotNullMark`. See the [implementation in the library](https://github.com/Kotlin/kotlinx.serialization/blob/master/core/commonMain/src/kotlinx/serialization/internal/NullableSerializer.kt#L17).

Reply:
The whole point here is, that the serializer being used is ("secretly") null-capable, so I'm expecting the downsteam serializer to call decodeNotNullMark and decodeNull. JsonNullSerializer does exactly that. This whole issue might be a bit JsonNull-specific I think.

Reply:
I misunderstood the meaning of this serializer. I thought it was the manually nullable serializer itself.